rpc/src/rpc.rs

```diff
@@ -1002,26 +1002,36 @@ impl JsonRpcRequestProcessor {
         })
     }
 
-    fn check_blockstore_root<T>(
+    // Check if the given `slot` is within the blockstore bounds. This function assumes that
+    // `result` is from a blockstore fetch, and that the fetch:
+    // 1) Checked if `slot` is above the lowest cleanup slot (and errored if not)
+    // 2) Checked if `slot` is a root
+    fn check_blockstore_bounds<T>(
         &self,
         result: &std::result::Result<T, BlockstoreError>,
         slot: Slot,
     ) -> Result<()> {
-        if let Err(err) = result {
-            debug!(
-                "check_blockstore_root, slot: {:?}, max root: {:?}, err: {:?}",
-                slot,
-                self.blockstore.max_root(),
-                err
-            );
-            if slot >= self.blockstore.max_root() {
-                return Err(RpcCustomError::BlockNotAvailable { slot }.into());
-            }
-            if self.blockstore.is_skipped(slot) {
-                return Err(RpcCustomError::SlotSkipped { slot }.into());
+        match result {
+            // The slot was found, all good
+            Ok(Some(_)) => Ok(()),
+            // The slot was cleaned up, return Ok() for now to allow fallback to bigtable
+            Err(BlockstoreError::SlotCleanedUp) => Ok(()),
+            // The slot was not cleaned up but also not found
+            Ok(_) => {
+                let max_root = self.blockstore.max_root();
+                debug!("check_blockstore_bounds, slot: {slot}, max root: {max_root}");
+                // Our node hasn't seen this slot yet, error out
+                if slot >= max_root {
+                    return Err(RpcCustomError::BlockNotAvailable { slot }.into());
+                }
+                // The slot is within the bounds of the blockstore as the lookup that yielded
+                // `result` checked that `slot` was greater than the blockstore's lowest
+                // cleanup slot and we just checked that `slot` was less than the blockstore's
+                // largest root. Thus, the slot must have been skipped and we can error out.
+                Err(RpcCustomError::SlotSkipped { slot }.into())
```

Comment:
I see why you make this assumption, but I think we would lose the disambiguation between a known skipped slot and some other sort of BlockstoreError. (Which would mostly be related to db corruption or inconsistency, or read failures, I think)

Reply:
> Maybe I'm missing some changes to the Blockstore get_rooted_block.. methods themselves that you're envisioning,

Nope, I just think I sufficiently confused myself with the `Option` things 🤦

> I think it's true (but please sanity check me) that if a slot is older than the latest_root but not cleaned up (as per Blockstore::check_lowest_cleanup_slot()) and not rooted (as per Blockstore::is_root()), it is definitely skipped (or missing due to ledger jump to recent snapshot, for which we return the same RPC error).

Yes, I believe this is to be the case and is a more succinct wording of what I was going for with this PR.
- The `Blockstore` functions that fetch the data of interest perform the slot-has-not-been-cleaned-up-yet check
- We still need the slot-is-older-than-latest-root check in `rpc.rs`
    - Could potentially move this check from `rpc` to `blockstore`, but let's keep a lid on that for now

Between `Blockstore::max_root()` and `Blockstore::is_skipped()`, the previous `check_blockstore_root()` function created 3 extra iterators (4 if you count the one in the `debug!()` statement):
- Two iterators for max root (one via `max_root()` and one via `is_skipped()`)
- One iterator for min root via `is_skipped()`

The end result of this PR is that we should be able to drop those 3 iterators to 1; we only need the extra `max_root()` call to distinguish skips from has-not-yet-occurred

Reply:
I liked your idea of storing max and min root on the Blockstore struct, btw, so we could eliminate all the iterators for root range use cases.

Reply:
> I liked your idea of storing max and min root on the Blockstore struct, btw, so we could eliminate all the iterators for root range use cases.

Cool, in order to keep things small, I think we can tackle that in a separate PR. This PR will minimize the number of root range checks, and then the subsequent PR will optimize the implementation of the root range checks. That work for you ?

Reply:
> That work for you ?

Yep, that sounds good!

Reply:
Yes, that issue is addressed! Resolving conversation.
Let me know when this PR is ready for formal review (after removal of `is_skipped()`?)